Mount all iso listed in bootstrap config file

ramanduh opened the pull request:
WIP from Nicolas's stash: https://github.com/scality/metalk8s/commit/c2aa4c538a0f2b8f3912887873e05899375a85a2
cf context for description

Mount all iso listed in /etc/metalk8s/bootstrap.yaml and configure salt-master

**Component**:

deployment

**Context**: 

Add a products entry in /etc/metalk8s/bootstrap.yaml which will contain
the product path or the product iso. If product path is provided,
expose saltstack file_roots and pillar_roots
into salt master configuration. If a iso image is provided, mount the iso
first before performing the same operation.

As set of assertion is also made to verify the product version. In this
case, if the products:metalk8s field contains an iso, isoinfo binary is
required to extract the product version

Change in the bootstrap config:

```
apiVersion: metalk8s.scality.com/v1alpha2
kind: BootstrapConfiguration
networks:
  controlPlane: 172.21.254.0/28
  workloadPlane: 172.21.254.32/27
ca:
  minion: bootstrap
apiServer:
  host: 172.21.254.14
  keepalived:
    enabled: true
products:
  metalk8s:
  - /srv/scality/metalk8s-2.0.0-dev/
```

**Note**
The PR is not fully functional right now. In fact, the `iso_root_path` pillar is moved to ext_pillar and generated by the salt master. The issue is that is the iso path provided by the user is not accessible from the salt master container, the ext_pillar will fail

**Summary**:

**Acceptance criteria**: 


---

See: #924

bert-e commented:
# Hello ramanduh,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/924-mount-all-iso-listed-in-bootstrap-config` and the
destination branch `development/2.0`.

Please resolve the conflict on **the feature branch** (`improvement/924-mount-all-iso-listed-in-bootstrap-config`).

```bash
 $ git fetch
 $ git checkout origin/improvement/924-mount-all-iso-listed-in-bootstrap-config
 $ git merge origin/development/2.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/924-mount-all-iso-listed-in-bootstrap-config
```

gdemonet commented on salt/_modules/metalk8s.py:
```diff
@@ -104,3 +110,43 @@ def minions_by_role(role, nodes=None):
     ]
+
+
+def _product_version_from_fd(fd):
```
I'd rather have this helper work with strings, much easier to test in isolation. Also, you can use `re.MULTILINE` to avoid testing all the lines by hand:

```python
def _get_product_version(info):
    match = re.search(r'^VERSION=(?P<version>.+)$', info, re.MULTILINE)
    return match.group('version') if match else None
```

ramanduh replied:
will fix

gdemonet commented on salt/_modules/metalk8s.py:
```diff
@@ -104,3 +110,43 @@ def minions_by_role(role, nodes=None):
+    product_txt = os.path.join(path, 'product.txt')
+
+    if not os.path.isfile(product_txt):
+        log.warning('Path %r has no "product.txt"', path)
```
Isn't this an error rather than a warning? I guess raising `CommandExecutionError` from here would make sense.

ramanduh replied:
will fix

gdemonet commented on salt/_modules/metalk8s.py:
```diff
@@ -104,3 +110,43 @@ def minions_by_role(role, nodes=None):
+    result = __salt__['cmd.run_all'](cmd=cmd)
+    log.debug('Result: %r', result)
+
+    if result['retcode'] != 0:
```
That would be an error as well IIUC

ramanduh replied:
will fix

gdemonet reviewed (commented):
Few light comments on the loading of `product.txt`

gdemonet commented:
Some stronger concerns:

- `products` seems rather vague, and since the current `BootstrapConfiguration` file is only for MetalK8s, I find it too generic (we wouldn't store Solution ISOs in this file, IIUC): we need to discuss this, as it's part of the user experience when installing MetalK8s. In any case, the `metalk8s.products` formula only works for MetalK8s ISOs (there are some `Assert product '{{ product }}' is MetalK8s` states).
- We allow either mounted or unmounted ISOs in this file, but it goes against the idea described in #924: we want to control where ISOs are mounted, so we can be sure we'll mount what is needed in the `salt-master`  and `repositories` containers.
- We now use the _full_ version: IMHO, it's out of this PR's scope, but if we want to do so, it will change a lot of documentation/UI w.r.t. the `saltenv` variable (we would need something like `saltenv=metalk8s-2.0.0-alpha1` instead of `saltenv=metalk8s-2.0`)
- We missed the `repositories` configuration (for the images), we only did `[file|pillar]_roots` for Salt master. This has to be part of this PR IMO, since the `metalk8s.products` formula is now part of the bootstrap script and upgrade procedure, as explained in #924
- [This comment](https://github.com/scality/metalk8s/issues/924#issuecomment-492202747) from @TeddyAndrieux also requests changes to the Salt master manifest to mount the accurate `[file|pillar]_roots`

TeddyAndrieux commented on salt/metalk8s/products/init.sls:
```diff
@@ -0,0 +1,81 @@
+    - file: Assert '{{ product_path }}/product.txt' exists
+
+{%   for (root, subdir) in [('file_roots', 'salt'), ('pillar_roots', 'pillar')] %}
+Expose Salt '{{ root }}' for {{ product_path }}:
```
We should directly call the salt-master state instead of doing it here

ramanduh replied:
Ok, will do

TeddyAndrieux commented on Vagrantfile:
```diff
@@ -132,10 +132,13 @@ apiServer:
     enabled: true
+products:
+  metalk8s:
+  - /srv/scality/metalk8s-$VERSION
```
No that's not what we want it's the oposite.

The path where to extract the iso should be hardcoded
We should have something like
```
products:
  metalk8s:
    - /root/metalk8s.iso
    - /var/tmp/metalk8s.iso
    - /toto/tata/titi.iso
```
And then the goal is to mount these iso in function of the version so for each `/srv/scality/metalk8s-<version>`

ramanduh replied:
Yes, it's already the case, the path where to mount the iso is hardcoded in `/srv/scality`, it is just that both are supported: product path and iso (if the iso is given, mount it otherwise, just config salt and perform a set of assert)

ramanduh commented:
@gdemonet thanks for reviews:
* Point 1: Ok for discussion about it. Is it only for the naming? (in that case, I will ping you directly) or do you want to discuss about the fact that we may want to manage solution iso as well (in that case, I will create an other question issue)
* Point 2: the mounted iso is allowed as the state will be bundled in the iso itself, so my understanding is that this state will be called necessary when at least one iso is mounted
* Point 3: I will see with @NicolasT using version instead short version is necessary or not before planing the doc/ui update
* Point 4 and 5: Ok, working on it

slaperche-scality commented on Vagrantfile:
```diff
@@ -80,8 +80,8 @@ die() {
 echo "Installing build artifacts on the host"
 
 mkdir -p /srv/scality || die "Failed to create /srv/scality"
-rm -f "/srv/scality/metalk8s-$SHORT_VERSION" || die "Failed to unlink symlink destination"
```
The move from `SHORT_VERSION` to `VERSION` clearly deserve a commit (and commit message) on its own.
(It's not even mentioned in the current one…)

ramanduh replied:
@gdemonet @slaperche-scality I'm gonna remove this change from this PR, you're right, it deserves a new issue/ticket

gdemonet commented:
> @gdemonet thanks for reviews:
> 
> * Point 1: Ok for discussion about it. Is it only for the naming? (in that case, I will ping you directly) or do you want to discuss about the fact that we may want to manage solution iso as well (in that case, I will create an other question issue)

No, the point is I don't think we want to manage **all** ISOs in this file, thus renaming/moving `products:metalk8s` to `archives`. Let's discuss it today.

> * Point 2: the mounted iso is allowed as the state will be bundled in the iso itself, so my understanding is that this state will be called necessary when at least one iso is mounted

See the original ticket (and comments): we don't wan't to let the user choose where ISOs will be mounted, even _the first one_. This could allow a user to mount its newest ISO at a random path, e.g. `/tmp/metalk8s-latest`, without any trouble w.r.t Pillar values (in particular, `metalk8s:iso_root_path`, which is hardcoded to `/srv/scality/metalk8s-@@VERSION`).

> * Point 3: I will see with @NicolasT using version instead short version is necessary or not before planing the doc/ui update

I don't see why it would be necessary. It's useful (because we'll be able to test upgrades between patch versions), but not mandatory in this specific PR.

ycointe commented on salt/metalk8s/products/init.sls:
```diff
@@ -0,0 +1,81 @@
+  - makedirs: true
+
+Product {{ product }} available at {{ product_path }}:
+  mount.mounted:
```
It would be nice to add `persist: True` to have the iso mounted automatically and to be resilient when a restart occur.

gdemonet replied:
Note that it's the default, but we want to be explicit

ramanduh replied:
fixed

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/924-mount-all-iso-listed-in-bootstrap-config` has diverged from
`development/2.0` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.0` into `improvement/924-mount-all-iso-listed-in-bootstrap-config`
* Rebase `improvement/924-mount-all-iso-listed-in-bootstrap-config` onto `origin/development/2.0`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config` with contents from `improvement/924-mount-all-iso-listed-in-bootstrap-config`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config origin/development/2.1
 $ git merge origin/improvement/924-mount-all-iso-listed-in-bootstrap-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config
```

Commit:
Install genisoimage on bootstrap

Relates #929

Commit:
No longer use iso_root_path get product from config file

Get the list of all directory/iso from the BootstrapConfig file through ext_pillar.
Use the ext_pillar value to create a dictionnary with saltenv and path,
used by the salt-master and repositories salt states.

See: #924

Commit:
Mount all iso listed in bootstrap config file

Add a products entry in /etc/metalk8s/bootstrap.yaml which will contain
the product path or the product iso. If product path is provided,
expose saltstack file_roots and pillar_roots
into salt master configuration. If a iso image is provided, mount the
iso
first before performing the same operation.

As set of assertion is also made to verify the product version. In this
case, if the products:metalk8s field contains an iso, isoinfo binary is
required to extract the product version

Relates: #924

Commit:
Allow to run bootstrap from different directory

Relates: #924

Commit:
Run boostrap role before minion role

As the part of mounting all iso is part of bootstrap

See #924

TeddyAndrieux commented on salt/_pillar/metalk8s.py:
```diff
@@ -78,12 +79,34 @@ def _load_apiserver(config_data):
+
+    for product in config['products']['metalk8s']:
+        if __salt__['file.directory_exists'](product):
+            version = __salt__['metalk8s.product_version_from_tree'](product)
```
Do we really want to support this ? For me we don't.
Any news about this @gdemonet or can I remove it ?

TeddyAndrieux replied:
Yes we need it, done

TeddyAndrieux commented on salt/metalk8s/products/mounted.sls:
```diff
@@ -0,0 +1,65 @@
+{% import_yaml "/etc/metalk8s/bootstrap.yaml" as bootstrap_config %}
```
Not really like this approach, strongly prefer using ext_pillar informations, anyone against that ?

TeddyAndrieux replied:
Fixed

TeddyAndrieux commented on salt/metalk8s/products/mounted.sls:
```diff
@@ -0,0 +1,65 @@
+{%     if not product_version %}
+{%         set product_path = product %}
+{%         set product_version = 'unknown' %}
+{%         set product_path_available = "- test: Product " ~ product ~ " available" %}
```
All these informations should be retrieved from ext_pillar

TeddyAndrieux commented on salt/_pillar/metalk8s.py:
```diff
@@ -78,12 +79,34 @@ def _load_apiserver(config_data):
+        if __salt__['file.directory_exists'](product):
+            version = __salt__['metalk8s.product_version_from_tree'](product)
+        else:
+            version = __salt__['metalk8s.product_version_from_iso'](product)
```
Ok we can't do that in ext_pillar as we need the iso to be accessible from the master if we want to do this

TeddyAndrieux replied:
Only retrieve list of iso/directory in ext_pillar, use them in a custom module

TeddyAndrieux committed:
Change CI bootstrap config according to this iso_root_path changes

Fixes: #924

TeddyAndrieux commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config` with contents from `improvement/924-mount-all-iso-listed-in-bootstrap-config`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config origin/development/2.1
 $ git merge origin/improvement/924-mount-all-iso-listed-in-bootstrap-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config
```

TeddyAndrieux commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config` with contents from `improvement/924-mount-all-iso-listed-in-bootstrap-config`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config origin/development/2.1
 $ git merge origin/improvement/924-mount-all-iso-listed-in-bootstrap-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.2/improvement/924-mount-all-iso-listed-in-bootstrap-config` with contents from `w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config`
and `development/2.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.2/improvement/924-mount-all-iso-listed-in-bootstrap-config origin/development/2.2
 $ git merge origin/w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.2/improvement/924-mount-all-iso-listed-in-bootstrap-config
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

bert-e commented:
# History mismatch



Merge commit #511e4ba2e416e90890c08d6422d045425c337dc8 on the integration branch
`w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config` is merging a branch which is neither the current
branch `improvement/924-mount-all-iso-listed-in-bootstrap-config` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/924-mount-all-iso-listed-in-bootstrap-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config` with contents from `improvement/924-mount-all-iso-listed-in-bootstrap-config`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config origin/development/2.1
 $ git merge origin/improvement/924-mount-all-iso-listed-in-bootstrap-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.2/improvement/924-mount-all-iso-listed-in-bootstrap-config` with contents from `w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config`
and `development/2.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.2/improvement/924-mount-all-iso-listed-in-bootstrap-config origin/development/2.2
 $ git merge origin/w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.2/improvement/924-mount-all-iso-listed-in-bootstrap-config
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

sayf-eddine-scality commented on salt/_modules/metalk8s.py:
```diff
@@ -104,3 +106,127 @@ def minions_by_role(role, nodes=None):
+
+
+def _get_product_version(info):
+    """Extract production version from info
```
product~ion~

TeddyAndrieux replied:
Dudu's fault :p

bert-e commented:
# History mismatch



Merge commit #ee6fc68e71520d8dc8443d35a280ef3550ffb356 on the integration branch
`w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config` is merging a branch which is neither the current
branch `improvement/924-mount-all-iso-listed-in-bootstrap-config` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/924-mount-all-iso-listed-in-bootstrap-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config` with contents from `improvement/924-mount-all-iso-listed-in-bootstrap-config`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config origin/development/2.1
 $ git merge origin/improvement/924-mount-all-iso-listed-in-bootstrap-config
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/924-mount-all-iso-listed-in-bootstrap-config
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

TeddyAndrieux committed:
Add `products` in `bootstrap.yaml` example in documentation

Fixes: #924

sayf-eddine-scality reviewed (approved):
LGTM
minor note: If we have two isos with the same mount folder the first is ignored without warning maybe we want to add one @TeddyAndrieux  ?

TeddyAndrieux commented:
> LGTM
> minor note: If we have two isos with the same mount folder the first is ignored without warning maybe we want to add one @TeddyAndrieux ?

Lets merge this one as it's here since long time and do it in another PR

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @alexandre-allard-scality

* @gdemonet

* @zempashi

* @sayf-eddine-scality

* @slaperche-scality

* @alxf

* @monpote

* @tcarmet

* @carlito-scality

* @ramanduh

* @cmonfort

* @thomasdanan

* @lucieleonard

* @nicolast

* @teddyandrieux

* @nootal

tcarmet commented:
@bert-e bypass_author_approval

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **bypass_author_approval**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue None.

Goodbye ramanduh.
